File: crypto-ffi/bindings/jvm/src/jmh/kotlin/com/wire/benchmark/RemoveUser.kt
@file:Suppress("ktlint:standard:no-wildcard-imports", "WildcardImport")

package com.wire.benchmark
import com.wire.crypto.*
import kotlinx.coroutines.runBlocking
import org.openjdk.jmh.annotations.*
import testutils.*
import java.nio.file.Files
import java.util.concurrent.TimeUnit

// This benchmark measures throughput of encrypting messages in a transaction and committing these.
// It includes the database interaction, because want to be able to answer to a user how long creating messages takes.
@BenchmarkMode(Mode.Throughput)
@OutputTimeUnit(TimeUnit.SECONDS)
@Warmup(iterations = 1, time = 1, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@State(Scope.Thread)
open class RemoveUser {
    @Param(
        "1",
        "2",
        "3",
        "5",
        "7"
    )
    var cipherSuite: UShort = 1u

    @Param("1", "10", "100")
    var userCount: Int = 0

    private lateinit var aliceCc: CoreCrypto
    private lateinit var conversationId: ConversationId
    private lateinit var clientIdsToRemove: List<ClientId>

    @Setup(Level.Invocation)
    fun setup() {
        runBlocking {
            val cipherSuite = CipherSuite.entries.first { it.value == cipherSuite }
            aliceCc = ccInit(CcInitOptions(CcInitOptions.Mode.WithBasicCredential(cipherSuite)))
            conversationId = createConversation(aliceCc)

            val keyPackages = mutableListOf<KeyPackage>()
            clientIdsToRemove = buildList {
                repeat(userCount) {
                    val bobId = genClientId()
                    val bobCc =
                        ccInit(
                            CcInitOptions(
                                mode = CcInitOptions.Mode.WithBasicCredential(cipherSuite),
                                clientId = bobId
                            )
                        )
                    val kp = generateKeyPackage(bobCc)
                    keyPackages.add(kp)
                    this.add(bobId)
                }
            }

            aliceCc.transaction {
                it.addClientsToConversation(conversationId, keyPackages)
            }
        }
    }

    @Benchmark
    fun bench() = runBlocking {
        aliceCc.transaction {
            it.removeClientsFromConversation(conversationId, clientIdsToRemove)
        }
    }
}

File: crypto-ffi/bindings/js/shared/benches/removeUser.ts
import type { ClientId, KeyPackage } from "#core-crypto-ffi";
import { runOnPlatform } from "#shared-utils";
import { userBenchmarkParameters } from "./utils";

export async function setupRemoveUserBench() {
    const parameters = await userBenchmarkParameters();

    await runOnPlatform(async (parameters) => {
        globalThis.bench = new tinybench.Bench({
            name: "Removing a User",
            time: 0,
            iterations: 10,
            warmup: true,
            warmupIterations: 1,
            warmupTime: 0,
            setup: globalThis.tinybenchSetup,
            teardown: globalThis.tinybenchTeardown,
        });
        for (const { userCount, cipherSuite } of parameters) {
            bench.add(
                `cipherSuite=${ccModule.CipherSuite[cipherSuite]} userCount=${userCount}`,
                async () => {
                    const aliceCc = await helpers.ccInit({
                        withBasicCredential: true,
                        cipherSuite,
                    });

                    const conversationId =
                        await helpers.createConversation(aliceCc);
                    const keyPackages: KeyPackage[] = [];
                    const clientIdsToRemove: ClientId[] = [];

                    for (let i = 0; i < userCount; i++) {
                        const bobId = helpers.newClientId();
                        const bobCc = await helpers.ccInit({
                            withBasicCredential: true,
                            cipherSuite,
                            clientId: bobId,
                        });
                        const kp = await helpers.generateKeyPackage(
                            bobCc,
                            cipherSuite
                        );
                        keyPackages.push(kp);
                        clientIdsToRemove.push(bobId);
                    }

                    await aliceCc.transaction(
                        async (ctx) =>
                            await ctx.addClientsToConversation(
                                conversationId,
                                keyPackages
                            )
                    );
                    const start = bench.now();

                    await aliceCc.transaction(
                        async (ctx) =>
                            await ctx.removeClientsFromConversation(
                                conversationId,
                                clientIdsToRemove
                            )
                    );

                    const end = bench.now();
                    return { overriddenDuration: end - start };
                }
            );
        }
    }, parameters);
}
